node/core/chain-selection/src/lib.rs

```diff
@@ -0,0 +1,574 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Implements the Chain Selection Subsystem.
+
+use polkadot_primitives::v1::{BlockNumber, Hash, Header, ConsensusLog};
+use polkadot_node_primitives::BlockWeight;
+use polkadot_subsystem::{
+	Subsystem, SubsystemContext, SubsystemError, SpawnedSubsystem,
+	OverseerSignal, FromOverseer,
+	messages::{ChainSelectionMessage, ChainApiMessage},
+	errors::ChainApiError,
+};
+
+use parity_scale_codec::Error as CodecError;
+use futures::channel::oneshot;
+use futures::prelude::*;
+
+use std::time::{UNIX_EPOCH, SystemTime};
+
+use crate::backend::{Backend, OverlayedBackend, BackendWriteOp};
+
+mod backend;
+mod tree;
+
+#[cfg(test)]
+mod tests;
+
+const LOG_TARGET: &str = "parachain::chain-selection";
+
+type Timestamp = u64;
+
+#[derive(Debug, Clone)]
+enum Approval {
+	// Approved
+	Approved,
+	// Unapproved but not stagnant
+	Unapproved,
+	// Unapproved and stagnant.
+	Stagnant,
+}
+
+impl Approval {
+	fn is_stagnant(&self) -> bool {
+		matches!(*self, Approval::Stagnant)
+	}
+}
+
+#[derive(Debug, Clone)]
+struct ViabilityCriteria {
+	// Whether this block has been explicitly reverted by one of its descendants.
+	explicitly_reverted: bool,
+	// The approval state of this block specifically.
+	approval: Approval,
+	// The earliest unviable ancestor - the hash of the earliest unfinalized
+	// block in the ancestry which is explicitly reverted or stagnant.
+	earliest_unviable_ancestor: Option<Hash>,
+}
+
+impl ViabilityCriteria {
+	fn is_viable(&self) -> bool {
+		self.is_parent_viable() && self.is_explicitly_viable()
+	}
+
+	// Whether the current block is explicitly viable.
+	// That is, whether the current block is neither reverted nor stagnant.
+	fn is_explicitly_viable(&self) -> bool {
+		!self.explicitly_reverted && !self.approval.is_stagnant()
+	}
+
+	// Whether the parent is viable. This assumes that the parent
+	// descends from the finalized chain.
+	fn is_parent_viable(&self) -> bool {
+		self.earliest_unviable_ancestor.is_none()
+	}
+}
+
+// Light entries describing leaves of the chain.
+//
+// These are ordered first by weight and then by block number.
+#[derive(Debug, Clone, PartialEq)]
+struct LeafEntry {
+	weight: BlockWeight,
+	block_number: BlockNumber,
+	block_hash: Hash,
+}
+
+impl PartialOrd for LeafEntry {
+	fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+		let ord = self.weight.cmp(&other.weight)
+				.then(self.block_number.cmp(&other.block_number));
+
+		if ord.is_ne() { Some(ord) } else { None }
```

Comment:
I'd suggest to use `!maches!(ord, std::cmp::Ordering::Equal)` if we want to merge it before our CI images are updated to rust 1.53.